python2-pybind11/challenge/containers.cpp

```diff
@@ -0,0 +1,62 @@
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "basics.hpp"
+#include "containers.hpp"
+
+namespace py = pybind11;
+
+PYBIND11_DECLARE_HOLDER_TYPE(T, std::shared_ptr<T>);
+
+namespace containers {
+
+class DoodadSetIterator {
+public:
+    /* Keep copy of DoodadSet elements (as shared_ptrs)
+     * This is needed because the iterator might outlive the container.
+     * A more elegant solution would be to have DoodadSet inherit from
+     * std::enable_shared_from_this<DoodadSet>, so that an
+     * iterator can keep the container alive while it is still in use
+     * even though the container has been garbage collected on the Python
+     * side, but that would require modifying the C++ implementation. */
```

Comment:
I still think this is quite possible; a Python iterator instance needs to hold a reference to its parent container to keep the parent from being destroyed, but the parent container never needs to hold a reference to its iterators, because it doesn't care when they're destroyed.  So if the iterator instance holds a reference to the Python container instance (which it does by increasing the reference count on the container when the iterator is created, and decrementing the reference count on the container when the iterator is destroyed).  It might be that `reference_internal` doesn't *quite* do that, so we might need a new policy, but it's similar enough that it should be at worst easy to add.